sysdump/sysdump.go

```diff
@@ -0,0 +1,825 @@
+				defer c.o.Logger.Debug("Finished collecting logs from Hubble Relay pods")
+				p, err := c.k.ListPods(ctx, c.o.HubbleRelayNamespace, metav1.ListOptions{
+					LabelSelector: c.o.HubbleRelayLabelSelector,
+				})
+				if err != nil {
+					return fmt.Errorf("failed to get logs from Hubble Relay pods")
+				}
+				if err := c.collectLogs(ctx, c.o.Logger, filterPods(p, nodeList), sinceTime, c.o.LogsLimitBytes, absoluteTempPath); err != nil {
+					return fmt.Errorf("failed to collect logs from Hubble Relay pods")
+				}
+				return nil
+			},
+			// Get logs from Hubble UI.
+			func() error {
+				c.o.Logger.Info("Collecting logs from Hubble UI pods")
+				defer c.o.Logger.Debug("Finished collecting logs from Hubble UI pods")
+				p, err := c.k.ListPods(ctx, c.o.HubbleNamespace, metav1.ListOptions{
+					LabelSelector: c.o.HubbleLabelSelector,
+				})
+				if err != nil {
+					return fmt.Errorf("failed to get logs from Hubble UI pods")
+				}
+				if err := c.collectLogs(ctx, c.o.Logger, filterPods(p, nodeList), sinceTime, c.o.LogsLimitBytes, absoluteTempPath); err != nil {
+					return fmt.Errorf("failed to collect logs from Hubble UI pods")
+				}
+				return nil
+			},
+		}...)
+	}
+
+	// Run the tasks.
+	errCh := make(chan error)
+	wgdCh := make(chan struct{})
+	var wg sync.WaitGroup
+	wg.Add(len(tasks))
+	for _, f := range tasks {
+		go func(f func() error) {
+			defer wg.Done()
+			if err := f(); err != nil {
+				errCh <- err
+			}
+		}(f)
+	}
+	go func() {
+		wg.Wait()
+		close(wgdCh)
+	}()
+
+	// Wait for the tasks to finish.
+	var merr error
+loop:
+	for {
+		select {
+		case <-wgdCh:
+			break loop
+		case err := <-errCh:
+			merr = multierror.Append(merr, err)
+		}
+	}
+
+	// Check if any errors occurred and warn the user.
+	if merr != nil {
+		c.o.Logger.Warnf("The sysdump may be incomplete — %s", merr.Error())
+		c.o.Logger.Warnf("Please note that depending on your Cilium version and installation options, this may be expected")
+	}
+
+	// Create the zip file in the current directory.
+	c.o.Logger.Info("Compiling sysdump")
+	p, err := os.Getwd()
+	if err != nil {
+		return fmt.Errorf("failed to get current directory: %w", err)
+	}
+	f := filepath.Join(p, replaceTimestamp(c.o.OutputFileName)+".zip")
+	if err := archiver.Archive([]string{d}, f); err != nil {
+		return fmt.Errorf("failed to create zip file: %w", err)
+	}
+	c.o.Logger.Infof("The sysdump has been saved to %s", f)
+
+	// Try to remove the temporary directory.
+	c.o.Logger.Debugf("Removing the temporary directory %s", d)
+	if err := os.RemoveAll(d); err != nil {
+		c.o.Logger.Warnf("failed to remove temporary directory %s: %v", d, err)
+	}
+	return nil
+}
+
+func buildNodeNameList(n *corev1.NodeList, filter string) ([]string, error) {
+	w := strings.Split(strings.TrimSpace(filter), ",")
+	r := make([]string, 0)
+loop:
+	// Iterate nodes and try to match either names or IPs to any of the specified filter strings.
+	for _, n := range n.Items {
```

Comment:
Could we maybe extract the inner loop logic into a dedicated function for better readability?